CollateralAccount/CollateralAccount.md

```diff
@@ -0,0 +1,106 @@
+---
+CIP: ?
+Title: Collateral Account for HD Wallets
+Authors: Sebastien Guillemot <seba@dcspark.io>
+Comments-URI: https://forum.cardano.org/t/collateral-account-derivation/65879
+Status: Draft
+Type: Standards
+Created: 2021-06-29
+License: CC-BY-4.0
+---
+
+# Abstract
+
+This document describes using a separate derivation path to ensure there is always sufficient collateral to execute UTXO smart contracts
+
+# Motivation
+
+Collateral input | As of Alonzo, transactions that call non-native smart contracts are required to put up collateral to cover the potential cost of smart contract execution failure. Inputs used as collateral has the following properties:
+
+- Cannot contain any tokens (only ADA)
+- Cannot be a script address
+- Must be a UTXO input
+- Must be at least some percentage of the fee in the tx (concrete percentage decided by a protocol parameter)
+- Can be the same UTXO entry as used in regular tx input
+- Is consumed entirely (no change) if the contract execution fails during phase 2 validation
+
+We therefore need a way for the wallet to handle picking which inputs to use for collateral
+
+## Naive solution
+
+The naive solution would do the following steps
+1. Add any regular input that satisfies the constraints as collateral as well
+1. If collateral is still insufficient, look at the wallet UTXO for any other entry that can be added as collateral
```

Comment:
Later in this CIP, if I understood correctly, it is implied that light wallets have no way to know how much of collateral is actually needed so they have to select all utxo in collateral account and at best provide some guesses of how much of collateral would be needed - which then means here in the naive solution the step of accessing whether the selected collateral is sufficient or not is impossible at all.

Reply:
According to the specs collateral is a function of the tx fee, so it can be computed beforehand.

Reply:
Collateral is not an explicit field in transactions -- it's calculated based on the tx fee which depends on the amount you include as part of the witness for each redeemer script.

Reply:
Ah, but the problem then is that the correct fee would depend on the execution units then as well, so if I'm getting it right, rather than knowing the correct collateral from knowing the correct fee, a wallet that has no access to a local personal full-node would not be able to correctly get neither the fee, not the collateral 🤔

Reply:
Yes, that's correct. That's why there is talk about trying to get the PAB to run in the browser so it can be included as part of light wallets. Otherwise the best you can do is trust some 3rd party to tell the truth about the cost of running the contract and risk losing collateral if they lied.

Reply:
There are two independent issues:

1. A wallet does need some information from the chain to be able to construct a transaction. It needs to know the UTxOs that make up the user's wallet so it can do coin selection. It needs to know the value of a few protocol parameters etc. Based on that it can work out the executions and fees for the transaction it builds. The hash of the relevant parameters (price of execution units etc) is included in the tx body and is checked in phase one, and thus no collateral can be lost.

2. Collateral needs to be supplied to cover the fees. Note that these are the fees that are actually declared. So again, if you get the fees wrong then the transaction is simply rejected and no collateral is taken.

The only way to arrange to have your collateral taken is to calculate the execution units incorrectly, but to include the correct protocol parameters into the transaction body. That is you cannot get this wrong by 3rd parties lying to you about current protocol parameters. You have to do something really wrong.



> Ah, but the problem then is that the correct fee would depend on the execution units then as well,
Yes.
> so if I'm getting it right, rather than knowing the correct collateral from knowing the correct fee, a wallet that has no access to a local personal full-node would not be able to correctly get neither the fee, nor the collateral
A wallet needs _some_ information from the chain like the wallet's UTxO, and the values of some of the current protocol parameters. If you're thinking about a light wallet and a compromised server lying to you, then there's no problem: because the hash of the relevant protocol parameters is included in the tx body and checked by the ledger rules in phase 1, you will not loose collateral if you were given the wrong protocol parameter values, or if the values changed before the tx is submitted.

> Yes, that's correct. That's why there is talk about trying to get the PAB to run in the browser so it can be included as part of light wallets. Otherwise the best you can do is trust some 3rd party to tell the truth about the cost of running the contract and risk losing collateral if they lied.

No. Unless there's a design mistke here that I am not aware of then I don't think you can lose collateral if the server lies. Provided that your local wallet can run the scripts locally then the only reason to loose collateral is bugs in the wallet. You do not need the whole PAB to be able to run scripts. There are more lightweight libraries that can do that. For example the CLI will do this using the Cardano API library.

Reply:
@dcoutts what other lightweight libraries can do that? by "Cardano API" you mean `cardano-wallet`?

> Provided that your local wallet can run the scripts locally

I must be missing something but I'm not sure how a pure JS wallet like Yoroi will be able to do this. If a light wallet needs to rely on a remote node every time it needs to recompute the fees and re-validate a smart-contract tx, this doesn't translates into a great UX.

I'm also confused because you mention that the wallet only needs to know its UTXO set and some protocol parameters in order to compute the execution units, but then you mention that the wallet needs to run the scripts locally, which sound like a big additional requirement.

Reply:
>> so if I'm getting it right, rather than knowing the correct collateral from knowing the correct fee, a wallet that has no access to a local personal full-node would not be able to correctly get neither the fee, nor the collateral
>
> A wallet needs some information from the chain like the wallet's UTxO, and the values of some of the current protocol parameters. If you're thinking about a light wallet and a compromised server lying to you, then there's no problem: because the hash of the relevant protocol parameters is included in the tx body and checked by the ledger rules in phase 1

@dcoutts , yes, but the main point is specifically about whether it is 100% guaranteed already that all wallets will be able to run scripts locally to validate their execution and calculate the execution points? 

> You do not need the whole PAB to be able to run scripts. There are more lightweight libraries that can do that. For example the CLI will do this using the Cardano API library.

Is this planned to be compatible with light wallets, e.g. allowing to run any script from any user, without having an instance of a wallet created locally?